Missing deps(?): spack install tr gives fatal error: X11/Xlib.h: No such file or directory

# Problem
`tr` comes with variant `X` and it's off by default:
```sh
Variants:
    Name Default   Description

    X    off       Enable X11 support
```
I get a build error when running (*):
```sh
$ spack install tk
[...]
$ more /tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/spack-build.out
[...]
$ 
/tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
```
suggesting that Tk still needs to `#include <X11/Xlib.h>` when being built (full details below).

BTW, I noticed this error when trying to install R as `spack install r`, which in turn depends on `tk` meaning the above error was a showstopper.  With my below workaround Tk and R installs.


# Troubleshooting
However,  [`repos/builtin/packages/tk/package.py`](https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/tk/package.py#L44) only includes `libx11` only when variant `X` is enabled (i.e. using `+X`):
```sh
    depends_on("tcl")
    depends_on("libx11", when='+X')
```

# Workaround
When removing `when='+X'` above (i.e. using `depends_on("libx11")`), then `spack install tk` installs Tk successfully.

Because of the error message, is the above a bug in `repos/builtin/packages/tk/package.py`?  I'm happy do a PR, but I'm not sure if this is the proper fix?

# Details
```sh
==> Installing tk
==> tcl is already installed in /netapp/home/hb/apps/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/tcl-8.6.6-knxyrpxcnki3akcjh2ppgfoufyl7r546
==> Fetching http://prdownloads.sourceforge.net/tcl/tk8.6.6-src.tar.gz
######################################################################## 100.0%
==> Staging archive: /netapp/home/hb/apps/spack/var/spack/stage/tk-8.6.6-oncwbwpqlkoloztntzehow7rad5pjfts/tk8.6.6-src.tar.gz
==> Created stage in /netapp/home/hb/apps/spack/var/spack/stage/tk-8.6.6-oncwbwpqlkoloztntzehow7rad5pjfts
==> Ran patch() for tk
==> Building tk [AutotoolsPackage]
==> Executing phase : 'autoreconf'
==> Executing phase : 'configure'
==> Executing phase : 'build'
==> Error: ProcessError: Command exited with status 2:
    'make'
/netapp/home/hb/apps/spack/lib/spack/spack/build_systems/autotools.py:275, in build:
     270      def build(self, spec, prefix):
     271          """Makes the build targets specified by
     272          :py:attr:``~.AutotoolsPackage.build_targets``
     273          """
     274          with working_dir(self.build_directory):
  >> 275              inspect.getmodule(self).make(*self.build_targets)

See build log for details:
/tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/spack-build.out
```

```sh
$ more /tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/spack-build.out
[...]
/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/ttk/ttkStubLib.c
In file included from /tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/tkPort.h:21:0,
                 from /tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/tkInt.h:19,
                 from /tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/tkStubLib.c:14:
/tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
In file included from /tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/ttk/ttkStubLib.c:9:0:
/tmp/hb/spack-stage/spack-stage-zTToCX/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
make: *** [tkStubLib.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [ttkStubLib.o] Error 1
```

# Session info
The above is on a fresh Centos 7:
```sh
rpm --query centos-release
centos-release-7-3.1611.el7.centos.x86_64
```
with today's version of Spack 0.10.0 (from `git clone https://github.com/llnl/spack.git`).

If you unpack Spack's  `r` package, `cd` into the staging directory and run `configure --help`, the output includes the following lines:

```
  --with-tcltk            use Tcl/Tk (if available), or specify its library
                          dir [yes]
  --with-tcl-config=TCL_CONFIG
                          specify location of tclConfig.sh []
  --with-tk-config=TK_CONFIG
                          specify location of tkConfig.sh []
```

The `tk8.6.6//README` describes Tk thusly:

> This directory contains the sources and documentation for Tk, an X11
toolkit implemented with the Tcl scripting language.

There are docs and knobs to turn that will allow one to build Tk w/ Aqua instead of X11 (on the mac).

`tk8.6.6/unix/configure --help` generates this:

```
hartzelg-UXG8WL:unix hartzelg$ ./configure --help
`configure' configures tk 8.6 to adapt to many kinds of systems.

Usage: ./configure [OPTION]... [VAR=VALUE]...

To assign environment variables (e.g., CC, CFLAGS...), specify them as
VAR=VALUE.  See below for descriptions of some of the useful variables.

Defaults for the options are specified in brackets.

Configuration:
  -h, --help              display this help and exit
      --help=short        display options specific to this package
      --help=recursive    display the short help of all the included packages
  -V, --version           display version information and exit
  -q, --quiet, --silent   do not print `checking...' messages
      --cache-file=FILE   cache test results in FILE [disabled]
  -C, --config-cache      alias for `--cache-file=config.cache'
  -n, --no-create         do not create output files
      --srcdir=DIR        find the sources in DIR [configure dir or `..']

Installation directories:
  --prefix=PREFIX         install architecture-independent files in PREFIX
			  [/usr/local]
  --exec-prefix=EPREFIX   install architecture-dependent files in EPREFIX
			  [PREFIX]

By default, `make install' will install all the files in
`/usr/local/bin', `/usr/local/lib' etc.  You can specify
an installation prefix other than `/usr/local' using `--prefix',
for instance `--prefix=$HOME'.

For better control, use the options below.

Fine tuning of the installation directories:
  --bindir=DIR           user executables [EPREFIX/bin]
  --sbindir=DIR          system admin executables [EPREFIX/sbin]
  --libexecdir=DIR       program executables [EPREFIX/libexec]
  --datadir=DIR          read-only architecture-independent data [PREFIX/share]
  --sysconfdir=DIR       read-only single-machine data [PREFIX/etc]
  --sharedstatedir=DIR   modifiable architecture-independent data [PREFIX/com]
  --localstatedir=DIR    modifiable single-machine data [PREFIX/var]
  --libdir=DIR           object code libraries [EPREFIX/lib]
  --includedir=DIR       C header files [PREFIX/include]
  --oldincludedir=DIR    C header files for non-gcc [/usr/include]
  --infodir=DIR          info documentation [PREFIX/info]
  --mandir=DIR           man documentation [PREFIX/man]

X features:
  --x-includes=DIR    X include files are in DIR
  --x-libraries=DIR   X library files are in DIR

Optional Features:
  --disable-FEATURE       do not include FEATURE (same as --enable-FEATURE=no)
  --enable-FEATURE[=ARG]  include FEATURE [ARG=yes]
  --enable-man-symlinks   use symlinks for the manpages (default: off)
  --enable-man-compression=PROG
                          compress the manpages with PROG (default: off)
  --enable-man-suffix=STRING
                          use STRING as a suffix to manpage file names
                          (default: no, tk if enabled without
                          specifying STRING)
  --enable-threads        build with threads (default: on)
  --enable-shared         build and link with shared libraries (default: on)
  --enable-64bit          enable 64bit support (default: off)
  --enable-64bit-vis      enable 64bit Sparc VIS support (default: off)
  --disable-rpath         disable rpath support (default: on)
  --enable-corefoundation use CoreFoundation API on MacOSX (default: on)
  --enable-load           allow dynamic loading and "load" command (default:
                          on)
  --enable-symbols        build with debugging symbols (default: off)
  --enable-aqua=yes|no    use Aqua windowingsystem on Mac OS X (default: no)
  --enable-xft            use freetype/fontconfig/xft (default: on)
  --enable-xss            use XScreenSaver for activity timer (default: on)
  --enable-framework      package shared libraries in MacOSX frameworks
                          (default: off)

Optional Packages:
  --with-PACKAGE[=ARG]    use PACKAGE [ARG=yes]
  --without-PACKAGE       do not use PACKAGE (same as --with-PACKAGE=no)
  --with-tcl              directory containing tcl configuration
                          (tclConfig.sh)
  --with-x                use the X Window System

Some influential environment variables:
  CC          C compiler command
  CFLAGS      C compiler flags
  LDFLAGS     linker flags, e.g. -L<lib dir> if you have libraries in a
              nonstandard directory <lib dir>
  CPPFLAGS    C/C++ preprocessor flags, e.g. -I<include dir> if you have
              headers in a nonstandard directory <include dir>
  CPP         C preprocessor

Use these variables to override the choices made by `configure' or to help
it to find libraries and programs with nonstandard names/locations.
```

There *is* a `--with-x` flag.  Out of curiousity, I ran `./configure --without-x`  and the output included:

```
[...]
checking for X... disabled
checking for X11 header files... couldn't find any!
checking for X11 libraries... checking for XCreateWindow in -lXwindow... no
could not find any!  Using -lX11.
checking for main in -lXbsd... no
checking whether to use xft... no
checking for X11/XKBlib.h... no
checking whether XKeycodeToKeysym is deprecated... yes
checking whether to try to use XScreenSaver... yes
checking for X11/extensions/scrnsaver.h... no
checking for XScreenSaverQueryInfo... no
checking for XScreenSaverQueryInfo in -lXext... no
checking for XScreenSaverQueryInfo in -lXss... no
checking whether char is unsigned... no[...]
```

Even when I do that (and do some gross things to point it at a tcl install), running `make` sez:

```
fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
make: *** [tkStubLib.o] Error 1
```

Perhaps there *is* a way to have Tk w/out X on "unix", but I don't see it.

--- a/lib/spack/spack/cmd/flake8.py
+++ b/lib/spack/spack/cmd/flake8.py
@@ -59,7 +59,8 @@ exemptions = {
               r'^\s*version\(.*\)',
               r'^\s*variant\(.*\)',
               r'^\s*depends_on\(.*\)',
-              r'^\s*extends\(.*\)'],
+              r'^\s*extends\(.*\)',
+              r'^\s*patch\(.*\)'],
         # Exempt '@when' decorated functions from redefinition errors.
         811: [r'^\s*\@when\(.*\)'],
     },

--- a/var/spack/repos/builtin/packages/expect/expect_detect_tcl_private_header_os_x_mountain_lion.patch
+++ b/var/spack/repos/builtin/packages/expect/expect_detect_tcl_private_header_os_x_mountain_lion.patch
@@ -1,0 +1,23 @@
+Fix Tcl private header detection on macOS
+
+https://sourceforge.net/p/expect/patches/17/
+
+diff -Naur expect5.45.orig/tclconfig/tcl.m4 expect5.45/tclconfig/tcl.m4
+--- expect5.45.orig/tclconfig/tcl.m4	2010-11-09 11:42:10.000000000 -0800
++++ expect5.45/tclconfig/tcl.m4	2013-09-23 00:10:00.000000000 -0700
+@@ -3389,9 +3389,12 @@
+             # the framework's Headers and PrivateHeaders directories
+             case ${TCL_DEFS} in
+ 	    	*TCL_FRAMEWORK*)
+-		    if test -d "${TCL_BIN_DIR}/Headers" -a \
+-			    -d "${TCL_BIN_DIR}/PrivateHeaders"; then
+-			TCL_INCLUDES="-I\"${TCL_BIN_DIR}/Headers\" -I\"${TCL_BIN_DIR}/PrivateHeaders\" ${TCL_INCLUDES}"
++		    if test -d "${TCL_BIN_DIR}/Headers"; then
++			if test -d "${TCL_BIN_DIR}/PrivateHeaders"; then
++			    TCL_INCLUDES="-I\"${TCL_BIN_DIR}/Headers\" -I\"${TCL_BIN_DIR}/PrivateHeaders\" ${TCL_INCLUDES}"
++			elif test -d "${TCL_BIN_DIR}/Headers/tcl-private"; then
++			    TCL_INCLUDES="-I\"${TCL_BIN_DIR}/Headers\" -I\"${TCL_BIN_DIR}/Headers/tcl-private\" ${TCL_INCLUDES}"
++			fi
+ 		    else
+ 			TCL_INCLUDES="${TCL_INCLUDES} ${TCL_INCLUDE_SPEC} `echo "${TCL_INCLUDE_SPEC}" | sed -e 's/Headers/PrivateHeaders/'`"
+ 		    fi

--- a/var/spack/repos/builtin/packages/expect/package.py
+++ b/var/spack/repos/builtin/packages/expect/package.py
@@ -1,0 +1,79 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+import glob
+import os
+
+
+class Expect(AutotoolsPackage):
+    """Expect is a tool for automating interactive applications such as
+    telnet, ftp, passwd, fsck, rlogin, tip, etc."""
+
+    homepage = "http://expect.sourceforge.net/"
+    url      = "https://sourceforge.net/projects/expect/files/Expect/5.45/expect5.45.tar.gz/download"
+
+    version('5.45', '44e1a4f4c877e9ddc5a542dfa7ecc92b')
+
+    depends_on('tcl')
+
+    depends_on('automake', type='build')
+    depends_on('autoconf', type='build')
+    depends_on('libtool',  type='build')
+    depends_on('m4',       type='build')
+
+    force_autoreconf = True
+
+    patch('expect_detect_tcl_private_header_os_x_mountain_lion.patch', when='@5.45')
+
+    def configure_args(self):
+        spec = self.spec
+
+        args = [
+            # Without this, expect binary and library are not installed
+            '--exec-prefix={0}'.format(self.prefix),
+            '--enable-threads',
+            '--enable-shared',
+            '--enable-64bit',
+            '--with-tcl={0}'.format(spec['tcl'].prefix.lib),
+            '--with-tclinclude={0}'.format(spec['tcl'].prefix.include),
+        ]
+
+        return args
+
+    @run_after('install')
+    def symlink_library(self):
+        """Expect installs libraries into:
+
+        lib/expect5.45/libexpect5.45.so
+
+        Create a symlink so that the library can be found in lib."""
+
+        target = join_path(self.prefix.lib, 'expect*', 'libexpect*')
+        target = glob.glob(target)[0]
+
+        link_name = os.path.basename(target)
+        link_name = join_path(self.prefix.lib, link_name)
+
+        symlink(target, link_name)

--- a/var/spack/repos/builtin/packages/tcl/package.py
+++ b/var/spack/repos/builtin/packages/tcl/package.py
@@ -49,7 +49,7 @@ class Tcl(AutotoolsPackage):
         base_url = 'http://prdownloads.sourceforge.net/tcl'
         return '{0}/tcl{1}-src.tar.gz'.format(base_url, version)
 
-    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+    def setup_environment(self, spack_env, run_env):
         # When using Tkinter from within spack provided python+tk, python
         # will not be able to find Tcl/Tk unless TCL_LIBRARY is set.
         run_env.set('TCL_LIBRARY', join_path(self.prefix.lib, 'tcl{0}'.format(

--- a/var/spack/repos/builtin/packages/tcl/package.py
+++ b/var/spack/repos/builtin/packages/tcl/package.py
@@ -49,11 +49,18 @@ class Tcl(AutotoolsPackage):
         base_url = 'http://prdownloads.sourceforge.net/tcl'
         return '{0}/tcl{1}-src.tar.gz'.format(base_url, version)
 
-    def setup_environment(self, spack_env, run_env):
+    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
         # When using Tkinter from within spack provided python+tk, python
         # will not be able to find Tcl/Tk unless TCL_LIBRARY is set.
         run_env.set('TCL_LIBRARY', join_path(self.prefix.lib, 'tcl{0}'.format(
             self.spec.version.up_to(2))))
+
+    def install(self, spec, prefix):
+        with working_dir(self.build_directory):
+            make('install')
+
+            # Some applications like Expect require private Tcl headers.
+            make('install-private-headers')
 
     @run_after('install')
     def symlink_tclsh(self):

--- a/var/spack/repos/builtin/packages/tk/package.py
+++ b/var/spack/repos/builtin/packages/tk/package.py
@@ -38,10 +38,8 @@ class Tk(AutotoolsPackage):
     version('8.6.5', '11dbbd425c3e0201f20d6a51482ce6c4')
     version('8.6.3', '85ca4dbf4dcc19777fd456f6ee5d0221')
 
-    variant('X', default=False, description='Enable X11 support')
-
-    depends_on("tcl")
-    depends_on("libx11", when='+X')
+    depends_on('tcl')
+    depends_on('libx11')
 
     configure_directory = 'unix'
 
@@ -49,7 +47,7 @@ class Tk(AutotoolsPackage):
         base_url = "http://prdownloads.sourceforge.net/tcl"
         return "{0}/tk{1}-src.tar.gz".format(base_url, version)
 
-    def setup_environment(self, spack_env, run_env):
+    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
         # When using Tkinter from within spack provided python+tk, python
         # will not be able to find Tcl/Tk unless TK_LIBRARY is set.
         run_env.set('TK_LIBRARY', join_path(self.prefix.lib, 'tk{0}'.format(

--- a/var/spack/repos/builtin/packages/tk/package.py
+++ b/var/spack/repos/builtin/packages/tk/package.py
@@ -47,7 +47,7 @@ class Tk(AutotoolsPackage):
         base_url = "http://prdownloads.sourceforge.net/tcl"
         return "{0}/tk{1}-src.tar.gz".format(base_url, version)
 
-    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+    def setup_environment(self, spack_env, run_env):
         # When using Tkinter from within spack provided python+tk, python
         # will not be able to find Tcl/Tk unless TK_LIBRARY is set.
         run_env.set('TK_LIBRARY', join_path(self.prefix.lib, 'tk{0}'.format(

